components/leds.py

```diff
@@ -0,0 +1,403 @@
+        return (
+            self._morse_message.count(".")
+            + 3 * self._morse_message.count("-")
+            + 3 * self._morse_message.count(" ")
+        )
+
+    def _morse_calculation(self) -> Tuple[int, int, int]:
+        # Work out how far through the message we are
+        DOT_LENGTH = 0.15  # seconds
+        total_time = self._morse_length * DOT_LENGTH
+        elapsed_time = time.monotonic() - self.pattern_start_time
+        if elapsed_time > total_time:
+            self.set(DisplayType.RAINBOW)
+        running_total = 0.0
+        for token in self._morse_message:
+            if token == ".":
+                running_total += 1.0 * DOT_LENGTH
+            if token == "-" or token == " ":
+                running_total += 3.0 * DOT_LENGTH
+            if running_total > elapsed_time:
+                # This is the current character
+                if token == " ":
+                    return LedColours.OFF.value
+                else:
+                    return LedColours.BLUE.value
+        # Default - should never be hit
+        return LedColours.OFF.value
+
+    def choose_morse_message(self, _message=None) -> None:
+        # Choose a morse message at random, unless specific message requested
+        # Only use lowercase and spaces
+        MESSAGES = [
+            "KILL ALL HUMANS",
+            "MORSE CODE IS FOR NERDS",
+            "HONEYBADGER DONT CARE",
+            "GLHF",
+        ]
+        if _message is None:
+            message = random.choice(MESSAGES)
+        else:
+            message = _message.upper()
+        # Convert to dots and dashes
+        MORSE_CODE_DICT = {
+            "A": ".-",
+            "B": "-...",
+            "C": "-.-.",
+            "D": "-..",
+            "E": ".",
+            "F": "..-.",
+            "G": "--.",
+            "H": "....",
+            "I": "..",
+            "J": ".---",
+            "K": "-.-",
+            "L": ".-..",
+            "M": "--",
+            "N": "-.",
+            "O": "---",
+            "P": ".--.",
+            "Q": "--.-",
+            "R": ".-.",
+            "S": "...",
+            "T": "-",
+            "U": "..-",
+            "V": "...-",
+            "W": ".--",
+            "X": "-..-",
+            "Y": "-.--",
+            "Z": "--..",
+            "1": ".----",
+            "2": "..---",
+            "3": "...--",
+            "4": "....-",
+            "5": ".....",
+            "6": "-....",
+            "7": "--...",
+            "8": "---..",
+            "9": "----.",
+            "0": "-----",
+        }
+        self._morse_message = ""
+        for character in message:
+            if character != " ":
+                self._morse_message += " ".join(MORSE_CODE_DICT[character]) + "  "
+            else:
+                self._morse_message += "     "
+        # Add more space at the end
+        self._morse_message += "  "
+
+    def execute(self):
+        colour = self.calc_solid()
```

Comment:
What's this doing here? It looks like you're immediately overwriting this variable, discarding this calculation.